pkg/controller/registry/resolver/step_resolver.go

```diff
@@ -84,14 +85,122 @@ func NewOperatorStepResolver(lister operatorlister.OperatorLister, client versio
+type walkOption func(csv *v1alpha1.ClusterServiceVersion) error
+
+func WithCSVPhase(phase v1alpha1.ClusterServiceVersionPhase) walkOption {
+	return func(csv *v1alpha1.ClusterServiceVersion) error {
+		if csv == nil || csv.Status.Phase != phase {
+			return fmt.Errorf("clusterServiceVersion %s/%s in phase %s instead of %s", csv.GetNamespace(), csv.GetName(), csv.Status.Phase, phase)
+		}
+		return nil
+	}
+}
+
+func WithUniqueCSVs() walkOption {
+	visited := map[string]struct{}{}
+	return func(csv *v1alpha1.ClusterServiceVersion) error {
+		// Check if we have visited the CSV before
+		if _, ok := visited[csv.GetName()]; ok {
+			return fmt.Errorf("infinite replacement chain detected")
+		}
+
+		visited[csv.GetName()] = struct{}{}
+		return nil
+	}
+}
+
+// walkReplacementChain walks along the chain of clusterServiceVersions being replaced and returns
+// the last clusterServiceVersions in the replacement chain. An error is returned if any of the
+// clusterServiceVersions before the last is not in the replaces phase or if an infinite replacement
+// chain is detected.
+func WalkReplacementChain(csv *v1alpha1.ClusterServiceVersion, csvToReplacement map[string]*v1alpha1.ClusterServiceVersion, options ...walkOption) (*v1alpha1.ClusterServiceVersion, error) {
+	if csv == nil {
+		return nil, fmt.Errorf("clusterServiceVersion cannot be nil")
+	}
+
+	for {
+		// Check if there is a CSV that replaces this CSVs
+		next, ok := csvToReplacement[csv.GetName()]
+		if !ok {
+			break
+		}
+
+		// Check walk options
+		for _, o := range options {
+			if err := o(csv); err != nil {
+				return nil, err
+			}
+		}
+
+		// Move along replacement chain.
+		csv = next
+	}
+	return csv, nil
+}
+
+// isReplacementChainThatEndsInFailure returns true if the last CSV in the chain is in the failed phase and all other
+// CSVs are in the replacing phase.
+func isReplacementChainThatEndsInFailure(csv *v1alpha1.ClusterServiceVersion, csvToReplacement map[string]*v1alpha1.ClusterServiceVersion) (bool, error) {
+	lastCSV, err := WalkReplacementChain(csv, csvToReplacement, WithCSVPhase(v1alpha1.CSVPhaseReplacing), WithUniqueCSVs())
+	if err != nil {
+		return false, err
+	}
+	return (lastCSV != nil && lastCSV.Status.Phase == v1alpha1.CSVPhaseFailed), nil
+}
+
+// ReplacementMapping takes a list of CSVs and returns a map that maps a CSV's name to the CSV that replaces it.
+func ReplacementMapping(csvs []*v1alpha1.ClusterServiceVersion) map[string]*v1alpha1.ClusterServiceVersion {
+	replacementMapping := map[string]*v1alpha1.ClusterServiceVersion{}
+	for _, csv := range csvs {
+		if csv.Spec.Replaces != "" {
+			replacementMapping[csv.Spec.Replaces] = csv
+		}
+	}
+	return replacementMapping
+}
+
+func (r *OperatorStepResolver) cachePredicates(namespace string) ([]cache.Predicate, error) {
+	predicates := []cache.Predicate{}
+	csvs, err := r.csvLister.ClusterServiceVersions(namespace).List(labels.Everything())
```

Comment:
does it make more sense to filter the list by the copied-csv label here and only get back non-copied ones? Instead of checking below in the `for` loop whether or not it's copied.

Reply:
That seems reasonable to me from a readability standpoint.